tests/test_vqe.py

```diff
@@ -582,6 +716,139 @@ def test_passing_kwargs(self, coeffs, observables, expected):
+        """Test that an error is raised if observable optimization is requested outside of tape
+        mode."""
+        if qml.tape_mode_active():
+            pytest.skip("This test works with tape mode disabled")
+
+        dev = qml.device("default.qubit", wires=2)
+        hamiltonian = qml.vqe.Hamiltonian([1], [qml.PauliZ(0)])
+
+        with pytest.raises(ValueError, match="Observable optimization is only supported in tape"):
+            qml.VQECost(lambda params, **kwargs: None, hamiltonian, dev, optimize=True)
+
+    @pytest.mark.parametrize("interface", ["tf", "torch", "autograd"])
+    def test_optimize(self, interface, tf_support, torch_support):
+        """Test that a VQECost with observable optimization gives the same result as another
+        VQECost without observable optimization."""
+        if not qml.tape_mode_active():
+            pytest.skip("This test works with tape mode enabled")
+        if interface == "tf" and not tf_support:
+            pytest.skip("This test requires TensorFlow")
+        if interface == "torch" and not torch_support:
+            pytest.skip("This test requires Torch")
+
+        dev = qml.device("default.qubit", wires=4)
+        hamiltonian = big_hamiltonian
+
+        cost = qml.VQECost(
+            qml.templates.StronglyEntanglingLayers,
+            hamiltonian,
+            dev,
+            optimize=True,
+            interface=interface,
+        )
+        cost2 = qml.VQECost(
+            qml.templates.StronglyEntanglingLayers,
+            hamiltonian,
+            dev,
+            optimize=False,
+            interface=interface,
+        )
+
+        w = qml.init.strong_ent_layers_uniform(2, 4, seed=1967)
+
+        c1 = cost(w)
+        exec_opt = dev.num_executions
+        dev._num_executions = 0
+
+        c2 = cost2(w)
+        exec_no_opt = dev.num_executions
+
+        assert exec_opt == 5
+        assert exec_no_opt == 15
+
+        assert np.allclose(c1, c2)
+
+    def test_optimize_grad(self):
+        """Test that the gradient of VQECost is accessible and correct when using observable
+        optimization and the autograd interface."""
+        if not qml.tape_mode_active():
+            pytest.skip("This test works with tape mode enabled")
+
+        dev = qml.device("default.qubit", wires=4)
+        hamiltonian = big_hamiltonian
+
+        cost = qml.VQECost(qml.templates.StronglyEntanglingLayers, hamiltonian, dev, optimize=True)
+        cost2 = qml.VQECost(
+            qml.templates.StronglyEntanglingLayers, hamiltonian, dev, optimize=False
+        )
+
+        w = qml.init.strong_ent_layers_uniform(2, 4, seed=1967)
+
+        dc = qml.grad(cost)(w)
+        exec_opt = dev.num_executions
+        dev._num_executions = 0
+
+        dc2 = qml.grad(cost2)(w)
+        exec_no_opt = dev.num_executions
+
+        assert exec_no_opt > exec_opt
+        assert np.allclose(dc, big_hamiltonian_grad)
+        assert np.allclose(dc2, big_hamiltonian_grad)
```

Comment:
The gradient tests are currently failing but should pass once https://github.com/PennyLaneAI/pennylane/pull/901 is merged in.